src/streams.c

```diff
@@ -765,99 +765,47 @@ Obj FuncPrint (
     return 0;
 }
 
-static Obj PRINT_OR_APPEND_TO(Obj args, int append)
+static Obj PRINT_OR_APPEND_TO_FILE_OR_STREAM(Obj args, int append, int file)
 {
     const char * volatile funcname = append ? "AppendTo" : "PrintTo";
     volatile Obj        arg;
-    volatile Obj        filename;
+    volatile Obj        destination;
     volatile UInt       i;
     syJmp_buf           readJmpError;
 
-    /* first entry is the filename                                         */
-    filename = ELM_LIST(args,1);
-    RequireStringRep(funcname, filename);
+    /* first entry is the file or stream                                   */
+    destination = ELM_LIST(args,1);
 
-    /* try to open the file for output                                     */
-    i = append ? OpenAppend( CONST_CSTR_STRING(filename) )
-               : OpenOutput( CONST_CSTR_STRING(filename) );
-    if ( ! i ) {
-        if (strcmp(CONST_CSTR_STRING(filename), "*errout*") == 0) {
-            // When trying to print an error opening *errout* failed,
-            // We exit GAP after trying to print an error.
-            // First try printing an error to stderr
-            int ret = fputs("gap: panic, could not open *errout*!\n", stderr);
-            // If that failed, try printing to stdout
-            if(ret == EOF) {
-                fputs("gap: panic, could not open *errout*!\n", stdout);
-            }
-            SyExit(1);
-        }
-        ErrorQuit( "%s: cannot open '%g' for output",
-                   (Int)funcname, (Int)filename );
-        return 0;
+    /* try to open the output                                              */
+    if ( file ) {
+        RequireStringRep(funcname, destination);
+        i = append ? OpenAppend( CONST_CSTR_STRING(destination) )
+                   : OpenOutput( CONST_CSTR_STRING(destination) );
     }
+    else
+        i = OpenOutputStream(destination);
 
-    /* print all the arguments, take care of strings and functions         */
-    for ( i = 2;  i <= LEN_PLIST(args);  i++ ) {
-        arg = ELM_LIST(args,i);
-        if ( IS_PLIST(arg) && 0 < LEN_PLIST(arg) && IsStringConv(arg) ) {
-            PrintString1(arg);
-        }
-        else if ( IS_STRING_REP(arg) ) {
-            PrintString1(arg);
-        }
-        else if ( TNUM_OBJ(arg) == T_FUNCTION ) {
-            PrintFunction( arg );
+    /* handle failures                                                     */
+    if ( ! i ) {
+        if ( file ) {
+            if (strcmp(CSTR_STRING(destination), "*errout*") == 0) {
+                Panic("Failed to open *errout*!");
+            }
+            ErrorQuit( "%s: cannot open '%g' for output",
+                       (Int)funcname, (Int)destination );
```

Comment:
I think I'd find it easier to follow the code if the error handling was move up, to directly after the `Open*` calls, so that you don't have another `if (file) ... else ...` here.